src/Data/Algorand/Transaction.hs

```diff
@@ -120,53 +119,41 @@ data OnComplete
   | OnCompleteClearState
   | OnCompleteUpdateApplication
   | OnCompleteDeleteApplication
-  deriving (Eq, Show)
+  deriving (Bounded, Eq, Enum)
 
 instance Default OnComplete where
   def = OnCompleteNoOp
 instance CanonicalZero OnComplete
 instance NonZeroValue OnComplete
 
-instance Enum OnComplete where
-  fromEnum OnCompleteNoOp = 0
-  fromEnum OnCompleteOptIn = 1
-  fromEnum OnCompleteCloseOut = 2
-  fromEnum OnCompleteClearState = 3
-  fromEnum OnCompleteUpdateApplication = 4
-  fromEnum OnCompleteDeleteApplication = 5
-
-  toEnum 0 = OnCompleteNoOp
-  toEnum 1 = OnCompleteOptIn
-  toEnum 2 = OnCompleteCloseOut
-  toEnum 3 = OnCompleteClearState
-  toEnum 4 = OnCompleteUpdateApplication
-  toEnum 5 = OnCompleteDeleteApplication
-  toEnum x = error ("Expect `OnComplete` integer value to be from 0 - 5 but got: " <> show x)
+instance Show OnComplete where
+  show OnCompleteNoOp = "noop"
+  show OnCompleteOptIn = "optin"
+  show OnCompleteCloseOut = "closeout"
+  show OnCompleteClearState = "clear"
+  show OnCompleteUpdateApplication = "update"
+  show OnCompleteDeleteApplication = "delete"
+
+instance Read OnComplete where
+  readsPrec _ = \case
+    "noop" -> [(OnCompleteNoOp, "")]
+    "optin" -> [(OnCompleteOptIn, "")]
+    "closeout" -> [(OnCompleteCloseOut, "")]
+    "clear" -> [(OnCompleteClearState, "")]
+    "update" -> [(OnCompleteUpdateApplication, "")]
+    "delete" -> [(OnCompleteDeleteApplication, "")]
+    x -> error ("Unmatch `OnComplete` string value. Got: " <> show x)
 
 instance AlgoMessagePack OnComplete where
-  toAlgoObject = toAlgoObject @Word64 . fromIntegral . fromEnum
-  fromAlgoObject o = (toEnum . fromIntegral) <$> fromAlgoObject @Word64 o
+  toAlgoObject = toAlgoObject @Word8 . fromIntegral . fromEnum
+  fromAlgoObject o = (toEnum . fromIntegral) <$> fromAlgoObject @Word8 o
 
 
 instance ToJSON OnComplete where
-  toJSON OnCompleteNoOp = toJSON @Text "noop"
-  toJSON OnCompleteOptIn = toJSON @Text "optin"
-  toJSON OnCompleteCloseOut = toJSON @Text "closeout"
-  toJSON OnCompleteClearState = toJSON @Text "clear"
-  toJSON OnCompleteUpdateApplication = toJSON @Text "update"
-  toJSON OnCompleteDeleteApplication = toJSON @Text "delete"
+  toJSON = toJSON . show
 
 instance FromJSON OnComplete where
-  parseJSON o = parseJSON @Text o
-    <&> (\a -> case a of
-      "noop" -> OnCompleteNoOp
-      "optin" -> OnCompleteOptIn
-      "closeout" -> OnCompleteCloseOut
-      "clear" -> OnCompleteClearState
-      "update" -> OnCompleteUpdateApplication
-      "delete" -> OnCompleteDeleteApplication
-      x -> error ("Unmatch `OnComplete` string value. Got: " <> show x)
-    )
+  parseJSON o = read <$> parseJSON @String o
```

Comment:
as I remember - `read` is not really safe and it's better to keep your previous `FromJSON` definition

Reply:
@gromakovsky what do you think?

Reply:
Oh, right, `read` is partial and shouldn't be used. But you can use `readMaybe` or `readEither` for example.

Reply:
If `read` fail, don't we want to throw error? Using `readMaybe` or `readEither`, implies that we want to use some default value, which I think shouldn't be the case here.

Reply:
Alright, make sense. I've updated this now.